node/node.go

```diff
@@ -104,52 +108,56 @@ func New(conf *Config) (*Node, error) {
 		stop:          make(chan struct{}),
 		server:        &p2p.Server{Config: conf.P2P},
 		databases:     make(map[*closeTrackingDB]struct{}),
+		readOnly:      conf.ReadOnly,
 	}
 
 	// Register built-in APIs.
 	node.rpcAPIs = append(node.rpcAPIs, node.apis()...)
 
 	// Acquire the instance directory lock.
-	if err := node.openDataDir(); err != nil {
-		return nil, err
-	}
-	keyDir, isEphem, err := getKeyStoreDir(conf)
-	if err != nil {
-		return nil, err
+	if !node.readOnly {
+		if err := node.openDataDir(); err != nil {
+			return nil, err
+		}
+		keyDir, isEphem, err := getKeyStoreDir(conf)
+		if err != nil {
+			return nil, err
+		}
+		node.keyDir = keyDir
+		node.keyDirTemp = isEphem
 	}
-	node.keyDir = keyDir
-	node.keyDirTemp = isEphem
+
 	// Creates an empty AccountManager with no backends. Callers (e.g. cmd/geth)
 	// are required to add the backends later on.
 	node.accman = accounts.NewManager(&accounts.Config{InsecureUnlockAllowed: conf.InsecureUnlockAllowed})
```

Comment:
Can we try to keep all the `!node.readOnly` logic contained in fewer `if` blocks? It seems like we could merge above and below, if this assignment was moved up?